_rules/1520.md

```diff
@@ -1,21 +1,28 @@
 ---
 rule_id: 1520
 rule_category: maintainability
-title: Only use `var` when the type is very obvious
+title: Use `var` (implicit typing) and emphasize business purpose over mechanics
 severity: 1
 ---
-Only use `var` as the result of a LINQ query, or if the type is very obvious from the same statement and using it would improve readability. So don't
+In almost all cases we recommend the use of var, to emphasizes the semantic "business purpose" of the variable and downplay the "mechanical" details of its storage (matching Domain Driven Design ubiquitous language principles).
 
-	var item = 3;                              // what type? int? uint? float?
-	var myfoo = MyFactoryMethod.Create("arg"); // Not obvious what base-class or			
-	                                           // interface to expect. Also
-	                                           // difficult to refactor if you can't
-	                                           // search for the class
+    var amount = 2.1m;
+    var referral = new Referral("jane.smith@acme.com");
+    var discount = discountService.GetDiscountFromReferral(referral)
+    var discountedAmount = discount.ApplyTo(amount);
+    foreach(var discount in discountService.GetAll()) { ... }
+    using(var discountService = serviceProvider.GetDiscountService()) { ... }
```

Comment:
Here I often want to understand whether an interface is returned or a concrete class. This is not always important, but if I know there's an abstraction, I want to verify in a PR review that it's returning the abstraction.

Reply:
Sure, but that also depends on public or private applications and so on. A general rule could be that public methods should return objects as an interface, but at the same time when it is a Dto, (typically ending up serialized) then an interface is not necessary. In the end I would enforce interfaces where applicable at a different level, because even the interface could change without refactoring being necessary.

Reply:
How would you enforce that?

Reply:
Do you mean that only coding guidelines that can be enforced are appropriate? I'm not sure I understand the question. Are you implying that specifying an interface instead of using var is a way of enforcing this? 

The problem you are compensating for is that apparently you review code that uses a method, but you don't review the method itself. But isn't that the real place where this would be enforced? Any external dependency should be defined as an interface, and any object returning from that interface should itself be either an interface or a dto, as defined in the interface for the service that returns the object. That's the place to enforce something like that, just from reviewing the interface.

Reply:
> Any external dependency should be defined as an interface, and any object returning from that interface should itself be either an interface or a dto, as defined in the interface for the service that returns the object.

That is an opinion. A know a lot of folks that never use interfaces, unless they want to abstract a service (e.g. using role-based interfaces). 

> The problem you are compensating for is that apparently you review code that uses a method, but you don't review the method itself.

I prefer to be able to review a PR from inside Github, and then you may not be able to see the method involved. I _do_ sometimes pull down the entire PR to my local machine, but not always because it will hold up the review. So, if something is not immediately clear from the PR, then the code is not good enough for me.

Reply:
Reading this back, I guess I have the advantage that the team is and tries to be consistent. But if for you a PR is always completely clear from just the changes made, that is amazing to me. I often say that if you truly want to understand a PR, you have to look at the work item specifications, and then at the tests that were written for the PR, and then take it from there. Any other method quickly boils down to very superficial reviewing.

Reply:
It isn't as black and white as both you and me (from two years ago) said. Sometimes the PR is clear enough and contains everything I need and sometimes I need to pull down the PR and browse through the changes to see if it all makes sense.